react/features/screenshot-capture/worker.ts

```diff
@@ -1,29 +1,135 @@
+import pixelmatch from 'pixelmatch';
+
 import {
-    CLEAR_INTERVAL,
-    INTERVAL_TIMEOUT,
-    SET_INTERVAL
+    CLEAR_TIMEOUT,
+    MAX_FILE_SIZE,
+    PERCENTAGE_LOWER_BOUND,
+    SEND_CANVAS_DIMENSIONS,
+    SET_TIMEOUT,
+    TIMEOUT_TICK
 } from './constants';
 
-const code = `
-    var timer;
 
-    onmessage = function(request) {
-        switch (request.data.id) {
-        case ${SET_INTERVAL}: {
-            timer = setInterval(() => {
-                postMessage({ id: ${INTERVAL_TIMEOUT} });
-            }, request.data.timeMs);
-            break;
-        }
-        case ${CLEAR_INTERVAL}: {
-            if (timer) {
-                clearInterval(timer);
+let timer: ReturnType<typeof setTimeout>;
+let height: number;
+let width: number;
+let canvas: OffscreenCanvas;
+let ctx: OffscreenCanvasRenderingContext2D | null;
+let storedImageData: ImageData | null;
+
+/**
+ * Sends Blob with the screenshot to main thread.
+ *
+  * @param {ImageData} imageData - The image of the screenshot.
+ * @returns {void}
+ */
+async function sendBlob(imageData: ImageData) {
+    let imageBlob = await canvas.convertToBlob({ type: 'image/jpeg' });
+
+    if (imageBlob.size > MAX_FILE_SIZE) {
+        const quality = Number((MAX_FILE_SIZE / imageBlob.size).toFixed(2)) * 0.92;
+
+        imageBlob = await canvas.convertToBlob({ type: 'image/jpeg',
+            quality });
+    }
+
+    storedImageData = imageData;
+
+    postMessage({
+        id: TIMEOUT_TICK,
+        imageBlob
+    });
+}
+
+/**
+ * Sends empty message to main thread.
+ *
+ * @returns {void}
+ */
+function sendEmpty() {
+    postMessage({
+        id: TIMEOUT_TICK
+    });
+}
+
+/**
+ * Draws the image bitmap on the canvas and checks the difference percent with the previous image
+ * if there is no previous image the percentage is not calculated.
+ *
+ * @param {ImageBitmap} imageBitmap - The image bitmap that is drawn on canvas.
+ * @returns {void}
+ */
+function checkScreenshot(imageBitmap: ImageBitmap) {
+    ctx?.drawImage(imageBitmap, 0, 0, width, height);
+    const imageData = ctx?.getImageData(0, 0, width, height);
+
+    imageBitmap.close();
+
+    if (!imageData) {
+        sendEmpty();
+
+        return;
+    }
+
+    if (!storedImageData) {
+        sendBlob(imageData);
+
+        return;
+    }
+
+    let numOfPixels = 0;
+
+    try {
+        numOfPixels = pixelmatch(
+            imageData.data,
+            storedImageData.data,
+            null,
+            width,
+            height);
+    } catch {
+        sendEmpty();
+
+        return;
+    }
+
+    const percent = numOfPixels / imageData.data.length * 100;
+
+    if (percent >= PERCENTAGE_LOWER_BOUND) {
+        sendBlob(imageData);
+    } else {
+        sendEmpty();
+    }
+
+}
+
+onmessage = function(request) {
+    switch (request.data.id) {
+    case SEND_CANVAS_DIMENSIONS: {
+        width = request.data.width;
+        height = request.data.height;
+        canvas = new OffscreenCanvas(width, height);
```

Comment:
Why don't you initialize that just once at the start of the script?

Reply:
the offscreen canvas requires the width and height arguments, this is why I send the width and height of the stream, to init it with the right dimensions, so that the image resulted from the blob has the correct size

Reply:
and I init it every time in case you stop sharing 2 screens having different dimensions

Reply:
but indeed I could just set it's width and height

Reply:
Let's do that, I think it will make things simpler.